File: code/BTree.hs
-- CPSC 312 - 2017 - Binary Trees in Haskell
module BTree where

-- To run it, try:
-- ghci
-- :load BTree

-- a binary tree where the nodes are labelled with integers
data BTree = Empty
           | Node BTree Int BTree

-- tolist lst  retuens the list giving the inorder traversal of lst
tolist :: BTree -> [Int]
tolist Empty = []
tolist (Node lt val rt) =
     tolist lt ++ (val : tolist rt)

--- Example to try
-- tolist (Node (Node Empty 2 (Node Empty 4 Empty)) 5 (Node Empty 8 Empty))





-- tolist without append (++), using accumulators
tolista :: BTree -> [Int]
tolista lst =
    tolist2 lst []

-- tolist2 tree lst   returns the the list of elements of tree followed by the elements of lst
tolist2 :: BTree -> [Int] -> [Int]
tolist2 Empty acc = acc
tolist2 (Node lt val rt) acc =
     tolist2 lt  (val : tolist2 rt acc)

--- Example to try
-- tolista (Node (Node Empty 2 (Node Empty 4 Empty)) 5 (Node Empty 8 Empty))

File: code/btree.pl
% CPSC 312 2017, Binary trees

% Tree to list (with append)
tolist(empty,[]).
tolist(node(LT,Val,RT),L):-
    tolist(LT,L0),
    tolist(RT,L1),
    append(L0,[Val|L1],L).

% Test case:
% tolist(node(node(empty, 2, node(empty, 4, empty)), 5, node(empty, 8, empty)),L).






% Tree to list (without append)
tolist_dl(T,L) :-
    tolist3(T,L,[]).

% tolist3(T, L1, L2) is true if L1-L2 is preorder list of values in tree T
tolist3(empty, L, L).
tolist3(node(LT,Val,RT),L0,L2):-
    tolist3(LT,L0,[Val|L1]),
    tolist3(RT,L1,L2).

% tolist_dl(node(node(empty, 2, node(empty, 4, empty)), 5, node(empty, 8, empty)),L).
